### ISSUE TITLE
TLS version (and/or cipher suite) negotiation appears to be broken with TLS 1.3 (discovered via Chromium GREASE).

### ISSUE THREAD

**Issue by cloudstrife9999** [pallets/flask#4984]
If Flask is configured to use TLS1.3+ (see code example below), the implementation appears to perform the TLS version negotiation (and/or the cipher suite negotiation) incorrectly.

Since TLS1.3 the version negotiation has been moved to the `supported_versions` extension in ClientHello/ServerHello.

In particular, the client sends a **discrete** list of supported versions in the ClientHello, and the server selects one that it supports, echoing it back to the client in the ServerHello.

In order to detect implementation bugs, Chromium sends random reserved (see [RFC 8701](https://www.rfc-editor.org/rfc/rfc8701.html)) version codes together with the legitimate version codes in the ClientHello. This is part of the "GREASE" mechanism to detect future "intolerances" in advance.

The correct approach when receiving "GREASE" codes / extensions is to simply ignore them as unrecognised codes / extensions.

It looks like that, when at least one of the "GREASE" version codes is greater than the TLS 1.3 code (0x0304), Flask aborts the TLS handshake with an alert saying that no common TLS versions / TLS cipher suites have been found.

Actually, it is possible that the error is caused by the same kind of "intolerance" in the cipher suite negotiation.
However, since "GREASE" is controlled by the browser, there is no easy way to check which of the two scenarios (version negotiation, and cipher suite negotiation) is causing the issue (or if both are).

Conversely, Mozilla Firefox does not send "GREASE" codes/extensions, and, as a result, the TLS handshake is successful.

The following code snippet can be used to trigger the issue with Chromium as the client.

```python
#!/usr/bin/env python3

from flask.app import Flask
from ssl import SSLContext, PROTOCOL_TLS_SERVER


app: Flask = Flask(__name__)

context: SSLContext = SSLContext(protocol=PROTOCOL_TLS_SERVER)

# Forcing TLS 1.3.
context.minimum_version = context.maximum_version

# Override with the actual values (this is just an example).
server_cert_path: str = "PUT_THE_CERTIFICATE_PATH_HERE"
server_private_key_path: str = "PUT_THE_PRIVATE_KEY_PATH_HERE"

context.load_cert_chain(certfile=server_cert_path, keyfile=server_private_key_path)

@ app.route("/")
def index() -> str:
    return "Hello World!"

app.run(host="127.0.0.1", port=5000, debug=False, ssl_context=context)
```

Environment:

- Python version: 3.11.2
- Flask version: 2.2.3

**Comment by davidism** [comment]
The Werkzeug development server is a thin wrapper around Python's built in `http.server`. Neither `http.server` nor the Werkzeug development server are intended for production, and docs for both show many warnings about that. If you believe there is an issue with TLS handling, you should report it to the Python `http.server` maintainers.

**Comment by davidism** [comment]
Also note that both Python and Pallets have security policies and contact information for responsible disclosure. Please practice responsible communication about security issues rather than posting publicly.

**Comment by cloudstrife9999** [comment]
@davidism Thank you for your reply. I will report the bug to the `ssl` maintainer.

I am aware that Werkzeug is not intended for production.

With regard to opening a public issue, I did so, because it is not a security issue (rather a bug that prevents the TLS handshake from succeeding).

Finally, with regard to opening an issue here first, I did so because I did not know whether Flask adds its own TLS logic on the top of the `ssl`  module logic. Thanks to your reply, now I know.
